### PR TITLE
CFY-6091. Use postgres to get events

### PR DESCRIPTION
**Pull request by jcollado**
In this PR, the implementation of the `events` endpoints is updated to use a
SQL database as a backend instead of elasticsearch.

The responses send back to the client are compatible with the ones previously
sent when elasticsearch was the backend. Hence, no update is needed in the
clients.

Some comments and remaning work for future PRs:

- Some work on the client would be nice to avoid converting data to have the
  same format as was returned by elasticsearch.

- There's quite a bit validation code that should be refactored probably with
  the help of some library. `flask-restful` provides `reqparse`, but it's no
  longer maintained and the documentation points to `marshmallow` (`schematics`
  looks similar).

- The delete events endpoint still doesn't work.

- Events are now deleted on cascade with both deployments and executions, so
  the ability to delete events might be no longer needed.

- There's a `search` endpoint that still relies on elasticsearch. A similar
  update to the one in this PR is required before being able to remove
  elasticsearch as the log backend.

Requires: cloudify-cosmo/cloudify-manager-blueprints#455

### CODE REVIEW

File: rest-service/manager_rest/rest/resources.py [hunk 1]
```diff
@@ -803,13 +822,278 @@ def get(self, deployment_id, **kwargs):
...
+            A SQL query that returns the events found that match the conditions
+            passed as arguments.
+        :rtype: :class:`sqlalchemy.sql.elements.TextClause`
+
         """
-        List events for the provided Elasticsearch query
+        if _include is not None:
+            current_app.logger.error(
+                'Projections with `_include` parameter are not supported')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources.py]: In other request handlers a BadParametersError is raised which is then handled by the "exceptions_handled" decorator.. this is done in order to return a consistent JSON response in case there's an error with some known error codes. There are many examples for this in the code, just look for BadParametersError.

File: rest-service/manager_rest/rest/resources.py [hunk 2]
```diff
@@ -803,13 +822,278 @@ def get(self, deployment_id, **kwargs):
...
+
+        if not isinstance(filters, dict) or 'type' not in filters:
+            current_app.logger.error('Filter by type is expected')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources.py]: Same as previous comment.

File: rest-service/manager_rest/rest/resources.py [hunk 3]
```diff
@@ -803,13 +822,278 @@ def get(self, deployment_id, **kwargs):
...
+
+        if 'cloudify_event' not in filters['type']:
+            current_app.logger.error(
+                'At least `type=cloudify_event` filter is expected')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources.py]: Same as previous comment.

File: rest-service/manager_rest/rest/resources.py [hunk 4]
```diff
@@ -803,13 +822,278 @@ def get(self, deployment_id, **kwargs):
...
+                    Execution.id == bindparam('execution_id'),
+                )
+            )
+
+        if (not isinstance(sort, dict) or
+                '@timestamp' not in sort or
+                sort['@timestamp'] != 'asc'):
+            current_app.logger.error(
+                'Sorting ascending by `timestamp` is expected')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources.py]: ...

File: rest-service/manager_rest/rest/resources.py [hunk 5]
```diff
@@ -803,13 +822,278 @@ def get(self, deployment_id, **kwargs):
...
+        query = query.order_by('timestamp')
+
+        if not isinstance(pagination, dict):
+            current_app.logger.error(
+                'Expected `pagination` parameter')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources.py]: ...

File: rest-service/manager_rest/rest/resources.py [hunk 6]
```diff
@@ -803,13 +822,278 @@ def get(self, deployment_id, **kwargs):
...
+
+        if 'size' not in pagination:
+            current_app.logger.error(
+                'Expected `size` pagination parameter')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources.py]: ...

File: rest-service/manager_rest/rest/resources.py [hunk 7]
```diff
@@ -803,13 +822,278 @@ def get(self, deployment_id, **kwargs):
...
+
+        query = query.limit(bindparam('limit'))
+
+        if 'offset' not in pagination:
+            current_app.logger.error(
+                'Expected `offset` pagination parameter')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources.py]: ...

File: rest-service/manager_rest/rest/resources.py [hunk 8]
```diff
@@ -803,13 +822,278 @@ def get(self, deployment_id, **kwargs):
...
+        :type filters: dict(str, str)
+        :returns:
+            A SQL query that returns the number of events found that match the
+            conditions passed as arguments.
+        :rtype: :class:`sqlalchemy.sql.elements.TextClause`
+
+        """
+        if not isinstance(filters, dict) or 'type' not in filters:
+            current_app.logger.error('Filter by type is expected')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources.py]: ...

File: rest-service/manager_rest/rest/resources.py [hunk 9]
```diff
@@ -803,13 +822,278 @@ def get(self, deployment_id, **kwargs):
...
+
+        if 'cloudify_event' not in filters['type']:
+            current_app.logger.error(
+                'At least `type=cloudify_event` filter is expected')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources.py]: ...

File: rest-service/manager_rest/rest/resources_v2.py [hunk 10]
```diff
@@ -767,19 +801,43 @@ def post(self):
...
-                                  sort=sort,
-                                  range_filters=range_filters)
-        events = ManagerElasticsearch.search_events(body=query)
-        metadata = ManagerElasticsearch.build_list_result_metadata(query,
-                                                                   events)
-        ManagerElasticsearch.delete_events(events)
+        """Delete events/logs connected to a certain Deployment ID."""
+        if not isinstance(filters, dict) or 'type' not in filters:
+            current_app.logger.error('Filter by type is expected')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources_v2.py]: ...

File: rest-service/manager_rest/rest/resources_v2.py [hunk 11]
```diff
@@ -767,19 +801,43 @@ def post(self):
...
+
+        if 'cloudify_event' not in filters['type']:
+            current_app.logger.error(
+                'At least `type=cloudify_event` filter is expected')
+            return abort(400)
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources_v2.py]: ...

File: rest-service/manager_rest/storage/resource_models.py [hunk 12]
```diff
@@ -202,13 +202,20 @@ class Event(DerivedResource, DerivedMixin):
     }
 
     timestamp = db.Column(UTCDateTime, nullable=False, index=True)
+    deployment_fk = foreign_key(Deployment.storage_id, nullable=False)
```
Comment (idanmo) [inline rest-service/manager_rest/storage/resource_models.py]: Why is this needed if you can resolve the deployment id through the execution?

File: rest-service/manager_rest/storage/resource_models.py [hunk 13]
```diff
@@ -234,6 +241,7 @@ class Log(DerivedResource, DerivedMixin):
...
```
Comment (idanmo) [inline rest-service/manager_rest/storage/resource_models.py]: Same comment as for Event...

File: rest-service/manager_rest/rest/resources.py (hunk 1 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources.py (hunk 2 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources.py (hunk 3 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources.py (hunk 4 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources.py (hunk 5 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources.py (hunk 6 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources.py (hunk 7 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources.py (hunk 8 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources.py (hunk 9 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources_v2.py (hunk 10 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources_v2.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources_v2.py (hunk 11 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources_v2.py]: Fixed in 10bf726

File: rest-service/manager_rest/rest/resources_v2.py [hunk 14]
```diff
@@ -761,6 +757,18 @@ def get(self, _include=None, filters=None,
         :rtype: :class:`manager_rest.storage.storage_manager.ListResult`
 
         """
+        if 'execution_id' not in filters:
```
Comment (idanmo) [inline rest-service/manager_rest/rest/resources_v2.py]: I'm not sure about this validation as it would probably be a good idea to allow filtering according to deployment_id, blueprint_id etc.. but lets assume this is fine for now.

File: rest-service/manager_rest/storage/resource_models.py (hunk 12 above)
Comment (jcollado) [inline rest-service/manager_rest/storage/resource_models.py]: Fixed in 1f19c93

File: rest-service/manager_rest/storage/resource_models.py (hunk 13 above)
Comment (jcollado) [inline rest-service/manager_rest/storage/resource_models.py]: Fixed in 1f19c93

File: rest-service/manager_rest/rest/resources_v2.py (hunk 14 above)
Comment (jcollado) [inline rest-service/manager_rest/rest/resources_v2.py]: Yes, I agree. The goal in this PR was to support what the client needs and then implement missing features as required.
